### PR TITLE
[SYCLomatic] Support migration cudaOccupancyMaxActiveBlocksPerMultiprocessor

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 1]
```diff
@@ -620,6 +620,60 @@ class logical_group {
...
+/// \return Always return 0.
+inline int sycl_occupancy_max_active_wg_per_xecore(
+    int *num_wg, int wg_size, int slm_size = 0, int sg_size = 32,
+    bool used_barrier = false, bool used_large_grf = false) {
+  const int slm_size_per_xe_core = 64 * 1024;
+  const int max_barrier_registers = 32;
+  dpct::device_ext &dev = dpct::get_current_device();
+
+  size_t max_wg_size = dev.get_info<sycl::info::device::max_work_group_size>();
+  assert(wg_size <= max_wg_size && "Work-group size cannot exceed the limit.");
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp]: let's limit wg_size to max_wg_size, have return value to tell caller.

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 2]
```diff
@@ -620,6 +620,60 @@ class logical_group {
...
+/// Computes the active work-group number per Xe-Core when the occupancy is maximum.
+/// Ref to https://github.com/oneapi-src/oneAPI-samples/tree/master/Tools/GPU-Occupancy-Calculator
+/// \param [out] num_wg Active work-group number.
+/// \param [in] wg_size Work-group size.
+/// \param [in] slm_size Shared local memory size.
+/// \param [in] sg_size Sub-group size.
+/// \param [in] used_barrier Whether barrier is used.
+/// \param [in] used_large_grf Whether General Register File is used.
+/// \return Always return 0.
+inline int sycl_occupancy_max_active_wg_per_xecore(
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp]:
```suggestion
inline int calculate_max_active_wg_per_xecore
```

File: clang/test/dpct/helper_files_ref/include/util.hpp (hunk 1 above)
Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/util.hpp]: done

File: clang/test/dpct/helper_files_ref/include/util.hpp (hunk 2 above)
Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/util.hpp]: done

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 3]
```diff
@@ -382,6 +382,9 @@ DPCT_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::bits<ExperimentalFeatures>), Exp
         false),
              DPCT_OPT_ENUM("dpl-experimental-api", int(ExperimentalFeatures::Exp_DPLExperimentalAPI),
         "Experimental extension that allows oneDPL experimental API.\n",
+        false),
+             DPCT_OPT_ENUM("occupancy-calculation", int(ExperimentalFeatures::Exp_OccupancyCalculation),
+        "Experimental extension that allows occupancy calculation API.\n",
```
Comment (zhiweij1) [inline clang/include/clang/DPCT/DPCTOptions.inc]:
```suggestion
        "Experimental helper function used to calculate occupancy.\n",
```

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 4]
```diff
@@ -620,24 +620,33 @@ class logical_group {
     return _group_linear_range_in_parent;
   }
 };
-/// Computes the active work-group number per Xe-Core when the occupancy is maximum.
-/// Ref to https://github.com/oneapi-src/oneAPI-samples/tree/master/Tools/GPU-Occupancy-Calculator
+/// Computes the active work-group number per Xe-Core when the occupancy is
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp]:
```suggestion
/// Computes the max active work-group number per Xe-Core.
```

File: clang/include/clang/DPCT/DPCTOptions.inc (hunk 3 above)
Comment (zhiweij1) [reply clang/include/clang/DPCT/DPCTOptions.inc]: done

File: clang/test/dpct/helper_files_ref/include/util.hpp (hunk 4 above)
Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/util.hpp]: done

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**